src/lc3_vm/virtual_machine.rs

```diff
@@ -143,108 +158,267 @@ impl VM {
             }
             Opcode::ST { sr, offset } => {
-                println!("{sr}");
-                println!("{offset}");
+                // Get the word to store from the source register
+                let word = self
+                    .get_register_value(sr.into())
+                    .map_err(|err| VMError::Execute(format!("ST: {}", err)))?;
+
+                let pc_value = self
+                    .get_pc()
+                    .map_err(|err| VMError::Execute(format!("ST: {}", err)))?;
+                let offset = sign_extend_9_bits(offset);
+                // The address of the value is calculated by adding the incremented PC to the
+                // sign extended offset
+                let address = pc_value.wrapping_add(offset);
+                // Store the word into the calculated memory address
+                self.store_word(address, word)
+                    .map_err(|err| VMError::Execute(format!("ST: {}", err)))?;
             }
             Opcode::JSR { mode, offset } => {
-                println!("{mode}");
-                print!("{offset}");
+                // Save PC into R7
+                let pc_value = self
+                    .get_pc()
+                    .map_err(|err| VMError::Execute(format!("JSR: {}", err)))?;
+                self.update_register(7, pc_value)
+                    .map_err(|err| VMError::Execute(format!("JSR: {}", err)))?;
+                // Calculate offset depending on mode
+                let new_pc_value = if mode {
+                    // If the mode flag is set the PC is the sum of the incremented PC and the sign
+                    // extended offset
+                    pc_value.wrapping_add(sign_extend_11_bits(offset))
+                } else {
+                    // If the mode flag is not set the pc is the base register, we shift the value 6 times
+                    // to the right because the base address is stored in the 3 most significant bits of the offset
+                    self.get_register_value(offset >> 6)
+                        .map_err(|err| VMError::Execute(format!("JSR: {}", err)))?
+                };
+                // Jump PC
+                self.set_pc(new_pc_value)
+                    .map_err(|err| VMError::Execute(format!("JSR: {}", err)))?;
             }
             Opcode::AND { dr, sr1, mode, sr2 } => {
-                println!("{dr}");
-                println!("{sr1}");
-                println!("{mode}");
-                println!("{sr2}");
+                let source_register_1 = self
+                    .get_register_value(sr1.into())
+                    .map_err(|err| VMError::Execute(format!("AND {}", err)))?;
+                let rhs = if mode {
+                    // imm mode
+                    sign_extend_5_bits(sr2)
+                } else {
+                    self.get_register_value(sr2.into())
+                        .map_err(|err| VMError::Execute(format!("AND {}", err)))?
+                };
```

Comment:
Again, maybe it's cleaner if in the decoding phase you determine this and then have two `ADD` instructions with two different arguments (register and imm5). But the two ways are valid.

Reply:
Same as above